File: mfpt.jl
using Random
using LinearAlgebra
using Statistics
using PyPlot
using Formatting


## -----------------------------------------------------------------------------
## structures
## -----------------------------------------------------------------------------
Copy(x::T) where T = T([deepcopy(getfield(x, k)) for k ∈ fieldnames(T)]...)
mutable struct Cache
    q_12::Float64
    p_12::Float64
    p_12_hat::Float64
    q_old::Float64
    p_old::Float64
    V::Float64
    dVdx::Float64
    zeta::Float64
end
function Cache(; q_12::Float64=0.0,
                 p_12::Float64=0.0, p_12_hat::Float64=0.0,
                 q_old::Float64=0.0, p_old::Float64=0.0,
                 V::Float64=0.0, dVdx::Float64=0.0, zeta::Float64=0.0)
    return Cache(q_12, p_12, p_12_hat, q_old, p_old, V, dVdx, zeta)
end

mutable struct State
    q::Float64  ## position
    p::Float64  ## momentum
    t::Float64  ## time
end
function State(; q::Float64, p::Float64, t::Float64)
    return State(q, p, t)
end

mutable struct Params
    h::Float64
    tau::Float64
    gamma::Float64
    seed::Integer
    k::Float64
    m::Float64
    limit::Float64
end
function Params(; h::Float64, tau::Float64, gamma::Float64, seed::Integer,
                  k::Float64, m::Float64, limit::Float64)
    return Params(h, tau, gamma, seed, k, m, limit)
end

mutable struct Hist
    crosses::Int64
    times::Array{Float64,1}
    qhist::Array{Float64,1}
    phist::Array{Float64,1}
end
function Hist(; crosses::Int64, times::Array{Float64,1}, qhist::Array{Float64,1}, phist::Array{Float64,1})
    return Hist(crosses, times, qhist, phist)
end
## -----------------------------------------------------------------------------


## -----------------------------------------------------------------------------
## functions
## -----------------------------------------------------------------------------
function V(x, k)
    # return k .* (1.0 .- x.^2).^2
    return 0.5*k .* (x .+ 1.0).^2
end
function dVdx(x, k)
    # return (-4.0 * k) .* x .* (1.0 .- x.^2)
    return k .* (x .+ 1.0)
end
function d2Vdx2(out::Float64, x::Float64, k::Float64)
    # return (12.0 * k) .* (x.^2 .- 1.0/3.0)
    return k
end
function V!(out::Float64, x::Float64, k::Float64)
    # out = k .* (1.0 .- x.^2).^2
    out = 0.5*k .* (x .+ 1.0).^2
end
function dVdx!(out::Float64, x::Float64, k::Float64)
    # out = (-4.0 * k) .* x .* (1.0 .- x.^2)
    out = k .* (x .+ 1.0)
end
function d2Vdx2(x::Float64, k::Float64)
    return k
end
function d2Vdx2!(out::Float64, x::Float64, k::Float64)
    # out = (12.0 * k) .* (x.^2 .- 1.0/3.0)
    out = k
end
function hh(q, p, limit)
    return q - limit
    # return p - limit
end
function dhhdx(q, p, limit)
    return [1.0; 0.0]
end

function analytic(Sbar::State, Sstar::State, P::Params)
    ## setup
    qbar = Sbar.q
    qstar = Sstar.q
    pbar = Sbar.p
    pstar = Sstar.p
    Gstar = [dVdx(qstar, P.k); 0.0]
    gstar = dhhdx(qstar, pstar, P.limit)
    S = [1.0 0.0; 0.0 0.0]

    ## hessians
    Hbar = d2Vdx2(qbar, P.k)                  ## energy hess at x_eq0
    Hstar = d2Vdx2(qstar, P.k)                ## energy hess at x_eqc
    hstar = 0.0                               ## constraint hess at x_eqc

    ## calc B
    n = gstar ./ norm(gstar)
    n = reshape(n, length(n), 1)
    E = nullspace(Matrix(n'))
    B = det(E'*(Hstar - k*hstar)*E) * (norm(Gstar)^2 / det(E'*E))
    @assert(B > 0)

    ## calc rate
    pf = P.gamma .* Gstar'*S*Gstar
    pf *= sqrt(det(Hbar) / (2*pi*P.tau*abs(B)))
    estar = V(qstar, P.k)
    ebar = V(qbar, P.k)
    ef = exp(-( estar - ebar ) / P.tau)
    lam = pf * ef
    return [pf, ef, lam]
end
## -----------------------------------------------------------------------------


## -----------------------------------------------------------------------------
## integrate for fixed time horizon
## -----------------------------------------------------------------------------
function euler_fix(S::State, N::Int64; sigma::Float64, P::Params, C::Cache,
                                       max_cross::Int64=Int64(1e4))
    ## extract
    q = S.q; p = S.p; t = S.t; q_old = S.q; p_old = S.p
    h = P.h; gamma = P.gamma; k = P.k
    f = C.dVdx

    ## setup
    j::Int64 = 1
    crosses::Int64 = 0
    qhist = zeros(Float64, N)
    phist = zeros(Float64, N)
    times = zeros(Float64, max_cross)
    zetas = randn(N)
    etas = randn(N)
    qhist[1] = q
    phist[1] = p

    ## integrate
    for i = 2:N
        ## euler update
        # q += p .* h
        # p += h .* (-gamma .* p .- dVdx!(f, q, k)) .+ (sigma .* zetas[i])
        q += h .* (-gamma*dVdx!(f, q, k) .+ p) .+ (sigma .* etas[i])
        p += h .* (- dVdx!(f, q, k))

        ## check
        # if (q - limit)*(q_old - limit) < 0.0
        if (hh(q, p, limit) >= 0.0) && (hh(q_old, p_old, limit) < 0.0)
            j += 1
            crosses += 1
            times[j] = t + h
        end

        ## update
        q_old = q
        p_old = p
        t += h
        qhist[i] = q
        phist[i] = p
    end

    ## return
    H = Hist(crosses=crosses, qhist=qhist, times=diff(times[1:j]), phist=phist)
    return H
end
function integrator_fix(SS::State, N::Int64; PP::Params, CC::Cache, ut::Symbol,
                                             max_cross::Int64=Int64(1e4))
    """
    N: number of steps
    limit: q-threshold
    ut: update type option (:euler!, :baoab!)
    """
    ## setup
    S = Copy(SS)
    P = Copy(PP)
    Random.seed!(P.seed)

    ## sigma & integrate
    if ut == :euler!
        sigma = sqrt(2.0 * P.h * P.gamma * P.tau)
        H = euler_fix(S::State, N::Int64; sigma=sigma, P=P, C=CC, max_cross=max_cross)
    elseif ut == :baoab!
        sigma = sqrt(P.tau * (1.0 - exp(-2.0 * P.gamma * P.h)))
        H = baoab_fix(S::State, N::Int64; sigma=sigma, P=P, C=CC, max_cross=max_cross)
    elseif ut == :euler2!
        sigma = sqrt(2.0 * P.h * P.gamma * P.tau)
        H = euler2_fix(S::State, N::Int64; sigma=sigma, P=P, C=CC, max_cross=max_cross)
    elseif ut == :euler3!
        sigma = sqrt(2.0 * P.h * P.gamma * P.tau)
        H = euler3_fix(S::State, N::Int64; sigma=sigma, P=P, C=CC, max_cross=max_cross)
    end
    return H
end
## -----------------------------------------------------------------------------


## -----------------------------------------------------------------------------
## integrate for variable time horizon
## -----------------------------------------------------------------------------
function euler_fix_fail(S::State, N::Int64; sigma::Float64, P::Params, C::Cache)
    """
    run `N` steps of euler until failure
    """
    ## extract
    q = S.q; p = S.p; t = S.t; q_old = S.q
    h = P.h; gamma = P.gamma; k = P.k; limit = P.limit
    f = C.dVdx

    ## setup
    j::Int64 = 1
    fail::Int64 = 0
    qhist = zeros(Float64, N)
    phist = zeros(Float64, N)
    time::Float64 = 0.0
    zetas = randn(N)
    etas = randn(N)
    qhist[1] = q

    ## integrate
    for i = 2:N
        ## euler update
        # q += p .* h
        # p += h .* (-gamma .* p .- dVdx!(f, q, k)) .+ (sigma .* zetas[i])
        q += h .* (-gamma*dVdx!(f, q, k) .+ p) .+ (sigma .* etas[i])
        p += h .* (- dVdx!(f, q, k))

        ## check
        if hh(q, p, limit) >= 0.0
            fail = 1
            time = t + h
            qhist[i] = q
            phist[i] = p
            Sout = State(q=q, p=p, t=time)
            j = i
            break
        end

        ## update
        q_old = q
        t += h
        qhist[i] = q
        phist[i] = p
    end

    ## return
    Sout = State(q=q, p=p, t=time)
    return fail, Sout, qhist[1:j], phist[1:j]
end
function euler_var(S::State, N::Int64; sigma::Float64, P::Params, C::Cache, Nchunk::Int64=Int64(1e7))
    ## setup
    SS = Copy(S)
    crosses::Int64 = 0
    qhist = Float64[]
    phist = Float64[]
    times = Float64[]
    fail::Int64 = 0
    time::Float64 = 0.0

    ## integrate
    for i = 1:N
        while fail == 0
            fail, SS, histq, histp = euler_fix_fail(SS::State, Nchunk::Int64; sigma=sigma, P=P, C=C)
            push!(qhist, histq...)
            push!(phist, histp...)
            time = SS.t
        end
        SS = Copy(S)
        # SS.q = sigma*randn() - 1.0
        # SS.p = sigma*randn()
        push!(times, time)
        crosses += fail
        fail = 0
    end

    ## return
    H = Hist(crosses=crosses, qhist=qhist, phist=phist, times=times)
    return H
end
function integrator_var(SS::State, Nfail::Int64; PP::Params, CC::Cache, ut::Symbol)
    """
    Nfail: number of failures
    limit: q-threshold
    ut: update type option (:euler!, :baoab!)
    """
    ## setup
    S = Copy(SS)
    P = Copy(PP)
    Random.seed!(P.seed)

    ## sigma & integrate
    if ut == :euler!
        sigma = sqrt(2.0 * P.h * P.gamma * P.tau)
        H = euler_var(S::State, Nfail::Int64; sigma=sigma, P=P, C=CC)
    end
    return H
end

## -----------------------------------------------------------------------------
## diagnostics
## -----------------------------------------------------------------------------
function plot_diagnostics(H::Hist, P::Params; subsample=1000)
    ## histogram / density
    fig = figure(figsize=(10, 8))
    ax1 = fig[:add_subplot](3,1,1)
    n, bins, patches = ax1[:hist](H.qhist, 100, normed=1, label="numerical");
    println(patches)
    dd = bins[2] - bins[1]
    xmin = -2.0
    xmax = 2.0
    q = collect(range(xmin, stop=xmax, step=dd))

    PP = P.gamma / (2.0 * pi * P.tau) .* exp.(-V(q, P.k) / P.tau)
    Sbar = State(q=-1.0, p=0.0, t=0.0)
    Sstar = State(q=0.0, p=0.0, t=0.0)
    lam_ana = analytic(Sbar, Sstar, P)[3][1]
    # if P.gamma <= 0.01
    #     ## NOTE: not sure how to handle in underdamped case
    #     PP = P.gamma / (2.0 * pi * P.tau) * exp.(-V(q, P.k) / P.tau)
    #     lam_ana = P.gamma * exp(-abs(V(P.limit, P.k) - V(-1, P.k)) / P.tau) / (2.0 * pi * P.tau)
    # elseif (P.gamma > 0.01) && (P.gamma < 10.0)
    #     ## NOTE: prefactor gets washed out since constant
    #     PP = (2.0*pi)^(-1) * exp.(-V(q, P.k) / P.tau)
    #     lam_ana = exp(-abs(V(P.limit, P.k) - V(-1, P.k)) / P.tau) / (2.0 * pi)
    # elseif P.gamma >= 10.0
    #     PP = exp.(-V(q, P.k) / P.tau)
    #     lam_ana = exp(-abs(V(P.limit, P.k) - V(-1, P.k)) / P.tau) / (2.0 * pi * P.gamma)
    # end

    den = sum(PP) * dd
    ax1[:plot](q, PP ./ den, label="analytic")
    ax1[:set_xlim](xmin, xmax)
    legend()

    ## time series
    ax2 = fig[:add_subplot](3,1,2)
    ts = collect(range(P.h, stop=length(H.qhist)*P.h, step=P.h))[1:subsample:end]
    qs = H.qhist[1:subsample:end]
    ax2[:plot](ts, qs, label="position time-series")
    legend()

    ax3 = fig[:add_subplot](3,1,3)
    ax3[:hist](H.times, density=true)
    yscale("log", nonposy="clip")
    lam_dns = 1.0 / mean(H.times)
    x = collect(range(0, stop=maximum(H.times), length=1000))
    ax3[:plot](x, lam_ana .* exp.(-lam_ana .* x), label="analytic = $(round.(lam_ana, digits=5))")
    ax3[:plot](x, lam_dns .* exp.(-lam_dns .* x), label="dns = $(round.(lam_dns, digits=5))")
    legend()
    fig[:show]()
end
function compare_results(H::Hist, P::Params; S0::State)
    q0 = S0.q
    qt = P.limit
    Etau_ana = exp(abs(V(qt, P.k) - V(q0, P.k)) / P.tau)
    ## TODO: should have three separate analytic calcs depending on gamma?
    # lam_ana = exp(-abs(V(qt, P.k) - V(q0, P.k)) / P.tau) / (2.0 * pi * P.gamma)

    Sbar = State(q=-1.0, p=0.0, t=0.0)
    Sstar = State(q=0.0, p=0.0, t=0.0)
    lam_ana = analytic(Sbar, Sstar, P)[3][1]

    Etau_dns_1 = (H.crosses / length(H.qhist)) / P.h
    Etau_dns_2 = mean(H.times)
    lam_dns = 1.0 / Etau_dns_2

    println("crosses ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~")
    printfmt("DNS crosses                                          : {:d}\n", H.crosses)
    printfmt("DNS mean(time to cross) = (crosses / nsteps) / dt    : {:0.3e}\n\n", Etau_dns_1)

    println("exit time ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~")
    printfmt("DNS-time mean(time to cross)                         : {:0.3e}\n", Etau_dns_2)
    printfmt("ANA-time E[tau] = e^(|V(qt) - V(q0)| / tau)          : {:0.3e}\n", Etau_ana)
    printfmt("exit time ratio DNS/ANA                              : {:0.3e}\n\n", Etau_dns_2/Etau_ana)

    println("exit rate ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~")
    printfmt("DNS-rate mean(rate)                                  : {:0.3e}\n", lam_dns)
    printfmt("ANA-rate lam = 1/(2pi*gam) e^(|V(qt) - V(q0)| / tau) : {:0.3e}\n", lam_ana)
    printfmt("exit rate ratio DNS/ANA                              : {:0.3e}\n", lam_dns/lam_ana)
    println("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n")

    out = Dict()
    out[:Etau_dns_1] = Etau_dns_1
    out[:Etau_dns_2] = Etau_dns_2
    out[:Etau_ana] = Etau_ana
    out[:lam_dns] = lam_dns
    out[:lam_ana] = lam_ana
    return out
end
## -----------------------------------------------------------------------------

File: mfpt.py
import numpy as np
import scipy as sp
import matplotlib.pyplot as plt
import time


## functions
## -----------------------------------------------------------------------------
def V(x, k):
    return k*(1-x**2)**2

def dVdx(x, k):
    return -4*k*x*(1-x**2)

def feuler(x0, v0, nsteps,
           k=None, dt=None, sigma=None, gamma=None, threshold=None, seed=None):
    ## setup
    np.random.seed(seed)
    xnew = np.copy(x0)
    xold = np.copy(x0)
    v = np.copy(v0)
    zeta = np.random.normal(size=nsteps)
    t = 0.0
    cross = 0
    crosstimes = []
    xhist = []

    ## euler
    for i in range(nsteps):
        xnew = xold + v * dt
        v += dt*(-gamma*v - dVdx(xnew, k)) + sigma*zeta[i]
        t += dt
        if np.sign((xnew - threshold)*(xold - threshold)) < 0:
            cross += 1
            crosstimes += [t]
        xold = xnew
        xhist += [xnew]
    time_to_cross = np.diff(np.hstack([0, crosstimes]))
    return cross, time_to_cross, xhist

def feuler_single(x0, v0,
                  k=None, dt=None, sigma=None, gamma=None, threshold=None, seed=None):
    ## setup
    np.random.seed(seed)
    xnew = np.copy(x0)
    xold = np.copy(x0)
    v = np.copy(v0)
    t = 0.0
    cross = 0
    crosstimes = []
    xhist = []

    ## euler
    while cross == 0:
        xnew = xold + v * dt
        v += dt*(-gamma*v - dVdx(xnew, k)) + sigma*np.random.normal()
        t += dt
        if np.sign((xnew - threshold)*(xold - threshold)) < 0:
            cross += 1
            crosstimes += [t]
        xold = xnew
        xhist += [xnew]
    time_to_cross = np.diff(np.hstack([0, crosstimes]))
    return cross, time_to_cross, xhist

def feuler_multiple(x0, v0, nsims,
                    k=None, dt=None, sigma=None, gamma=None, threshold=None, seed=None):
    crosses = []
    times = []
    xhists = []
    for i in range(nsims):
        cross, time_to_cross, xhist = feuler_single(x0, v0,
                                                    k=k, dt=dt, sigma=sigma, gamma=gamma,
                                                    threshold=threshold, seed=seed*i)
        crosses += [cross]
        times += [time_to_cross]
        xhists += [xhist]
    crosses = np.hstack(crosses)
    times = np.hstack(times)
    xhists = np.hstack(xhists)
    return np.sum(crosses), times, xhists

def baoab(x0, v0, nsteps,
          k=None, dt=None, tau=None, gamma=None, threshold=None, seed=None):
    """ BAOAB from LM p. 271 with m = 1 """
    ## setup
    np.random.seed(seed)
    q = np.copy(x0)
    q_12 = np.copy(x0)
    q_new = np.copy(x0)
    p = np.copy(v0)
    p_12 = np.copy(v0)
    p_12_hat = np.copy(v0)
    p_new = np.copy(v0)
    h = np.copy(dt)
    zeta = np.random.normal(size=nsteps)
    sigma = np.sqrt(tau * (1-np.exp(-2*gamma*h)))
    t = 0.0
    cross = 0
    crosstimes = []
    xhist = []
    ## baoab
    for i in range(nsteps):
        p_12 = p - h/2 * dVdx(q, k)
        q_12 = q + h/2 * p_12
        p_12_hat = np.exp(-gamma*h) * p_12 + sigma*zeta[i]
        q_new = q_12 + h/2 * p_12_hat
        p_new = p_12_hat - h/2 * dVdx(q_new, k)
        t += dt
        if np.sign((q_new - threshold)*(q - threshold)) < 0:
            cross += 1
            crosstimes += [t]
        q = q_new
        p = p_new
        xhist += [q]
    time_to_cross = np.diff(np.hstack([0, crosstimes]))
    return cross, time_to_cross, xhist

def plot_diagnostics(xhist, k=None, tau=None, dt=None):
    ## histogram / density
    plt.figure(0)
    n, bins, patches = plt.hist(xhist, 100, normed=1)
    xmin = -2.0
    xmax = 2.0
    x = np.arange(xmin, xmax, bins[1]-bins[0])
    P = np.exp(-V(x, k) / tau)
    norm = np.sum(P)*(bins[1]-bins[0])
    plt.plot(x, P/norm)
    plt.xlim([xmin, xmax])

    ## time series
    plt.figure(1)
    plt.plot(np.arange(0, len(xhist)*dt, dt), xhist)
    plt.show()

def compare_results(cross, time_to_cross,
                    x0=None, xt=None, tau=None,
                    nsteps=None, k=None, gamma=None):
    Etau_ana = np.exp(np.abs(V(xt, k) - V(x0, k)) / tau)
    lam_ana = np.exp(-np.abs(V(xt, k) - V(x0, k)) / tau) / (2.0 * np.pi * gamma)
    print("DNS crosses                  : {:d}".format(cross))
    print("DNS crosses / nsteps         : {:0.3e}\n".format(float(cross) / float(nsteps*dt)))

    print("DNS-time mean(time to cross) : {:0.3e}".format(np.mean(time_to_cross)))
    print("ANA-time E[tau]              : {:0.3e}\n".format(Etau_ana))

    print("DNS-rate mean(rate)          : {:0.3e}".format(1.0/np.mean(time_to_cross)))
    print("ANA-rate lambda              : {:0.3e}\n".format(lam_ana))

    a = float(cross) / float(nsteps*dt)
    b = np.mean(time_to_cross)
    c = Etau_ana
    d = 1.0/np.mean(time_to_cross)
    e = lam_ana
    return a, b, c, d, e



## main
## -----------------------------------------------------------------------------
if __name__ == "__main__":
    ## params
    seed = 1234
    tau = 0.15
    tau = 0.35
    dt = 0.01
    k = 1.0
    gamma = 2.0
    sigma = np.sqrt(2.0*dt*gamma*tau)
    threshold = 0.0
    nsteps = int(1e6)
    nsims = 50
    x0 = -1.0
    v0 = 0.1
    xt = threshold
    T = int(dt * nsteps)

    ## one long chain
    ## euler
t0 = time.time()
cross, time_to_cross, xhist = feuler(x0, v0, nsteps,
                                  k=k, dt=dt, sigma=sigma, gamma=gamma,
                                  threshold=threshold, seed=seed)
t1 = time.time()
print(t1-t0)
plot_diagnostics(xhist, k=k, tau=tau, dt=dt)
a, b, c, d, e = compare_results(cross, time_to_cross, x0=x0, xt=xt, tau=tau, nsteps=nsteps, gamma=gamma, k=k)

    ## baoab
t0 = time.time()
cross, time_to_cross, xhist = baoab(x0, v0, nsteps,
                                  k=k, dt=dt, tau=tau, gamma=gamma,
                                  threshold=threshold, seed=seed)
t1 = time.time()
print(t1-t0)
plot_diagnostics(xhist, k=k, tau=tau, dt=dt)
a, b, c, d, e = compare_results(cross, time_to_cross, x0=x0, xt=xt, tau=tau, nsteps=nsteps, gamma=gamma, k=k)


    ## many shorter chains
    cross, time_to_cross, xhist = feuler_multiple(x0, v0, nsims,
                                  k=k, dt=dt, sigma=sigma, gamma=gamma,
                                  threshold=threshold, seed=seed)
    plot_diagnostics(xhist, k=k, tau=tau, dt=dt)
    a, b, c, d, e = compare_results(cross, time_to_cross,
                                    x0=x0, xt=xt, tau=tau, nsteps=len(xhist), gamma=gamma, k=k)

taus = [0.5, 0.4, 0.3, 0.25, 0.225, 0.2, 0.19, 0.18, 0.17, 0.16, 0.15, 0.145, 0.14, 0.135, 0.13]
aas = []
bs = []
cs = []
ds = []
es = []
for t in taus:
    print("temperature = {:0.3e}".format(t))
    print("------------------------------------")
    sigma = np.sqrt(2.0*dt*gamma*t)
    # cross, time_to_cross, xhist = feuler(x0, v0, nsteps,
    #                                   k=k, dt=dt, sigma=sigma, gamma=gamma,
    #                                   threshold=threshold, seed=seed)
    cross, time_to_cross, xhist = baoab(x0, v0, nsteps,
                                      k=k, dt=dt, tau=tau, gamma=gamma,
                                      threshold=threshold, seed=seed)
    # cross, time_to_cross, xhist = feuler_multiple(x0, v0, nsims,
    #                               k=k, dt=dt, sigma=sigma, gamma=gamma,
    #                               threshold=threshold, seed=seed)
    a, b, c, d, e = compare_results(cross, time_to_cross,
                                    x0=x0, xt=xt, tau=t, nsteps=len(xhist), gamma=gamma, k=k)
    aas += [a]
    bs += [b]
    cs += [c]
    ds += [d]
    es += [e]

plt.loglog(taus, ds, label="dns")
plt.loglog(taus, es, label="ana")
plt.plot(taus, [ds[i]/es[i] for i in range(len(taus))], label="d/e")
plt.legend()
